docs/glossary.rst

```diff
@@ -0,0 +1,274 @@
+        placement.
+        Ensemble operations (such as scatter, gather, broadcast, and
+        reduce), may be optimized within a single ensemble execution session.
+
+        The run time characteristics of an ensemble include the supported data
+        flow topology and the computing resources to allocate for ensemble
+        workers. When an ensemble scope is entered, the framework may collect
+        or launch new workers to support the ensemble work. Workers may continue
+        to receive additional tasks and data until the ensemble scope changes.
+        The scope of the ensemble session is thus also constrained by the
+        appropriateness of the allocated ensemble worker pool for the available
+        work.
+
+    framework
+        SCALE-MS is a set of specifications, support packages, and software
+        collaborations that provide a framework in which software tools are
+        executed. The term "framework" is used to refer abstractly to the
+        SCALE-MS software stack and to the facilities it coordinates,
+        particularly in documentation contexts where it is appropriate to avoid
+        details of software packaging or implementation.
+
+    function
+    operation implementation
+        A well defined computational element or data transformation that can be used
+        to add computational work to a graph managed by a Context. Inputs
+        are strongly specified, and behavior for a given set of inputs is deterministic
+        (within numerical stability). Outputs may not be well specified
+        until inputs are bound (*e.g.* until an instance is created).
+
+    graph state
+        The aggregation of state for operations (nodes) and data (edges),
+        constrained by directed
+        acyclic data flow topology. Granularity is not yet fully determined, but
+        state must account for completed and incomplete operations, and allow
+        distinction between idle and currently executing nodes.
+
+    operation
+    operation instance
+        A node in a work graph. Previously described as *WorkElement* or *node*.
+
+        Data sources and specific instances of operations are represented as
+        nodes in a work graph.
+
+        Identity: A node is the uniquely identifiable representation of a
+        :term:`function` instance, defined in terms of the inputs and
+        specified behavior.
+
+        Corollary: the definition of an operation (a node) is immutable once added to the graph.
+
+        Finer points:
+
+        * The outputs of an operation may be accessed and subscribed to at any time.
+        * Internally, operations may be stateful. They have metadata associated with
+          their degree of completion and, potentially, with references to other
+          resources to describe intermediate or final results.
+
+    port
+        Generic term for a named source, sink, resource, or binding hook on a node.
+
+    resource
```

Comment:
@mrshirts writes:
> I don't understand resource defined in terms of an API hook. A resource would be some sort of machine that can execute tasks. What is described here seems like something I would call (probably wrong words) as a "resource interface", or something like that?

Reply:
I believe there will be objects and interfaces as abstract as this definition, and the term appears in documentation snippets like "Operations can provide resources such as data Futures or opaque API handles..." I think a term at this level of abstraction is necessary. The questions are whether a different term is more appropriate and whether the definition is sufficient to disambiguate what is meant when the term appears in discussion.